### PR TITLE
Added improved .msh reader/writer + tests + example

### PR DESCRIPTION
**Pull request by vfonov**
Improved version of .msh reader/writer that supports loading triangle surface elements and tetrahedral volume elements, and
also physical tags associated with elements. Also, scalar and vector fields defined per node and per element are supported. 

Attached is a binary file for unit tests, to be placed in tests/data/ , result of FEM simulation of currents distribution inside a sphere. 
[MshTestData.zip](https://github.com/libigl/libigl/files/4984168/MshTestData.zip)

#### Check all that apply (change to `[x]`)
- [x] All changes meet [libigl style-guidelines](https://libigl.github.io/style-guidelines/).
- [x] Adds new .cpp file.
- [x] Adds corresponding unit test.
- [ ] This is a minor change.

### CODE REVIEW

**Comment by jdumas** [comment]
Cool! Could you submit a PR to [libigl-test-data](https://github.com/libigl/libigl-tests-data) with a *smaller* binary mesh to try the IO in the unit tests?

@qnzhou this looks like your msh loader/saver from PyMesh, maybe you could take a look at this PR.

**Comment by vfonov** [comment]
I changed the tests to use smaller .msh file ( see attached). 
[MshTestData.zip](https://github.com/libigl/libigl/files/5120831/MshTestData.zip)

**Comment by vfonov** [comment]
I created a pull request for test-data: https://github.com/libigl/libigl-tests-data/pull/6

File: include/igl/readMSH.cpp
Comment (qnzhou) [inline include/igl/readMSH.cpp:144]: Perhaps output a warning?

File: include/igl/writeMSH.cpp
Comment (qnzhou) [inline include/igl/writeMSH.cpp:19]: I think double underscore names are reserved for compiler's internal usage.  This helper function can be declared in anonymous namespace to avoid symbol leaking.

File: include/igl/writeMSH.cpp
Comment (qnzhou) [inline include/igl/writeMSH.cpp:59]: Not sure if lambda is more preferable over regular template functions even under c++14.

File: include/igl/writeMSH.cpp
Comment (qnzhou) [inline include/igl/writeMSH.cpp:150]: Output some warning will be helpful.

**Review by qnzhou — APPROVED** [review]
Thanks @vfonov !  Only minor comments.  I think at some point we should just refactor msh IO into a stand alone lib so other repo can just depend on it.

File: include/igl/readMSH.cpp
Comment (vfonov) [reply include/igl/readMSH.cpp:144]: ok, I added a message on stderr

File: include/igl/writeMSH.cpp
Comment (vfonov) [reply include/igl/writeMSH.cpp]: ok, using anonymous namespace.

File: include/igl/writeMSH.cpp
Comment (vfonov) [reply include/igl/writeMSH.cpp]: made it throw an error.

File: include/igl/writeMSH.cpp
Comment (jdumas) [reply include/igl/writeMSH.cpp]: hmm since libigl is a header-only library, maybe using a `namespace internal` might be preferable here.

Comment (qnzhou) [reply include/igl/writeMSH.cpp]: Good point, perhaps `internal` namespace is more preferable in this case.  Sorry.

Comment (vfonov) [reply include/igl/writeMSH.cpp]: so, 
```
namespace igl {
 namespace {
   template <typename T,typename Derived> 
    void __append_mat_to_vec(...
...
```
?

Comment (jdumas) [reply include/igl/writeMSH.cpp]:
```cpp
namespace igl {
namespace internal {

template<...>
void append_mat_to_vec() {
...
}

}
}
```

**Comment by jdumas** [comment]
Hi! So I was about to (finally) merge this, but as I tried out the tutorial entry, I was wondering if the shading is supposed to be like this?

Screen Shot 2020-09-20 at 11 50 03 PM

Wouldn't it be better with flat shading? Also, could you add a `.msh` file to [libigl/libigl-tutorial-data](https://github.com/libigl/libigl-tutorial-data) (different from the test data) so that the tutorial entry doesn't require command-line argument (and we're sure to have the correct data for the tutorial).

**Comment by vfonov** [comment]
yes, probably flat shading will be better. How do I enable that?

**Comment by jdumas** [comment]
Use `ViewerData::set_face_based(true)`

**Comment by vfonov** [comment]
Screenshot of the new version of tutorial code for reading .msh file:
image

**Review by jdumas — APPROVED** [review]
Ok I tried it and it looks good. I'll merge this PR once I get write access to merge https://github.com/libigl/libigl-tutorial-data/pull/1

**Pull request merged**
